packages/appium/lib/appium.js

```diff
@@ -241,6 +251,261 @@ class AppiumDriver extends DriverCore {
     }
   }
 
+  /**
+   * Initialize a new bidi connection
+   * @param {import('ws').WebSocket} ws The websocket connection object
+   * @param {import('http').IncomingMessage} req The connection pathname, which might include the session id
+   */
+  async onBidiConnection(ws, req) {
+    ws.on('error', (err) => {
+      // Can't do much with random errors on the connection other than log them
+      this.log.error(err);
+    });
+
+    try {
+      const pathname = req.url;
+      if (!pathname) {
+        throw new Error('Invalid connection request');
+      }
+      const bidiSessionRe = new RegExp(`${BIDI_BASE_PATH}/([^/]+)$`);
+      const bidiNoSessionRe = new RegExp(`${BIDI_BASE_PATH}/?$`);
+      const sessionMatch = bidiSessionRe.exec(pathname);
+      const noSessionMatch = bidiNoSessionRe.exec(pathname);
+
+      if (!sessionMatch && !noSessionMatch) {
+        throw new Error(
+          `Got websocket connection for path ${pathname} but didn't know what to do with it. ` +
+            `Ignoring and will close the connection`,
+        );
+      }
+
+      // Let's figure out which driver is going to handle this socket connection. It's either going
+      // to be a driver matching a session id appended to the bidi base path, or this umbrella driver
+      // (if no session id is included in the bidi connection request)
+
+      /** @type {import('@appium/types').ExternalDriver | AppiumDriver} */
+      let bidiHandlerDriver;
```

Comment:
this functions contains too much logic. Consider splitting it into smaller chunks

Reply:
i can try to chunk it out some more. the fact that we use closures to maintain references to the websocket objects makes it nice, so we don't have to store too many references to them on drivers etc...

Reply:
ok, i split it up into a number of methods.